paasta_tools/kubernetes/application/controller_wrappers.py

```diff
@@ -216,37 +217,31 @@ def create(self, kube_client: KubeClient) -> None:
         self.sync_horizontal_pod_autoscaler(kube_client)
 
     def deep_delete_and_create(self, kube_client: KubeClient) -> None:
-        self.deep_delete(kube_client)
-        timer = 0
-        while (
-            self.kube_deployment in set(list_all_deployments(kube_client))
-            and timer < 60
-        ):
-            sleep(1)
-            timer += 1
-
-        if timer >= 60 and self.kube_deployment in set(
-            list_all_deployments(kube_client)
-        ):
-            try:
-                force_delete_pods(
-                    self.item.metadata.name,
-                    self.kube_deployment.service,
-                    self.kube_deployment.instance,
-                    self.item.metadata.namespace,
-                    kube_client,
-                )
-            except ApiException as e:
-                if e.status == 404:
-                    # Deployment does not exist, nothing to delete but
-                    # we can consider this a success.
-                    self.logging.debug(
-                        "not deleting nonexistent deploy/{} from namespace/{}".format(
-                            self.kube_deployment.service, self.item.metadata.namespace
-                        )
+        # When deleting then immediately creating, we need to use Background deletion to ensure we can create the deployment immediately
+        self.deep_delete(kube_client, propagation_policy="Background")
+        # Also forcibly delete remaining pods
+        try:
+            force_delete_pods(
+                self.item.metadata.name,
+                self.kube_deployment.service,
+                self.kube_deployment.instance,
+                self.item.metadata.namespace,
+                kube_client,
+            )
+        except ApiException as e:
+            if e.status == 404:
+                # Deployment does not exist, nothing to delete but
+                # we can consider this a success.
+                self.logging.debug(
+                    "not deleting nonexistent deploy/{} from namespace/{}".format(
```

Comment:
This log line needs to change since the exception now only means that we failed to delete all pods, not that deployment deletion did not happen fast enough. That case is now in the check that you added below.

Reply:
I don't really understand this log line to begin with (or the comment in `force_delete_pods` indicating we needed to have deleted the deployment before hand) -- we are catching a 404 when attempting to delete pods by paasta.yelp.com/service and paasta.yelp.com/instance labels, rather than relating to the deployment object, so this would only throw a 404 if the pods were deleted.

However the effect is basically the same -- the pods probably got deleted before we got to this part, so we can consider it a success. I'll update to reflect this.

Reply:
Oh, well...now we see why I didn't see API requests for deleting pods when setting up these brutal deployments; we're invoking force_delete_pods incorrectly here, sending the deployment name as the service name. I will fix this bug as well.